### PR TITLE
MF-486 - Add provisioning command to CLI

### LINKED ISSUE #486: Add provisioning command to CLI
Add `provisioning` to CLI so that it can bulk-provision things and channels from CSV files.

### PR DESCRIPTION
**Pull request by drasko**
Resolves #486

Signed-off-by: drasko <drasko.draskovic@gmail.com>

Pull request title should be `MF-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](./CONTRIBUTING.md) file for detailed contributing guidelines.

### Which issue(s) does this PR fix/relate to?
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)

### CODE REVIEW

**Review by chombium — CHANGES_REQUESTED** [review]
Except the error handling for the file opening, everything looks good to me.

I'm wandering if we need some extra validation of the csv field values before calling the sdk

File: cli/provision.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+			c, err := os.Open(args[0])
+			if err != nil {
+				logError(err)
+				return
+			}
+			reader := csv.NewReader(bufio.NewReader(c))
+			for {
+				l, err := reader.Read()
+				if err == io.EOF {
```
Comment (dborovcanin) [inline cli/provision.go:128]: You can use switch over error here, the code would be little more readable.

File: cli/provision.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+					return
+				}
+
+				channels = append(channels, c)
+			}
+
+			flush(channels)
+		},
+	},
+	cobra.Command{
```
Comment (dborovcanin) [inline cli/provision.go:153]: I like this command.

**Review by dborovcanin — COMMENTED** [review]
I see that you've updated vendor. Should `Gopkg.toml` be updated? If not, please ignore this remark.

**Review by chombium — DISMISSED** [review]

File: cli/provision.go (hunk 2 above)
Comment (drasko) [reply cli/provision.go:153]: Thanks ;)

File: cli/provision.go (hunk 1 above)
Comment (drasko) [reply cli/provision.go:128]: Done. Looks readable.

**Review by dborovcanin — COMMENTED** [review]
I find this random name generator implementation... strange. I'd prefer a simple custom random string generator. I know that it's used only for testing, but anyway, I'd like to avoid adding a new unnecessary dependency. Other than that, this PR looks OK.

File: cli/provision.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+				if err != nil {
+					logError(err)
+					return
+				}
+
+				things = append(things, m)
+			}
+			// Create channels
+			for i := 0; i < nbChan; i++ {
+				n := "c" + strconv.Itoa(i)
```
Comment (anovakovic01) [inline cli/provision.go]: Use `fmt.Sprintf`.

File: cli/provision.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+				l, err := reader.Read()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					logError(err)
+					return
+				}
+
+				m, err := createThing(l[0], l[1], args[1])
```
Comment (anovakovic01) [inline cli/provision.go:95]: Check for length of `l`.

File: cli/provision.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+	},
+	cobra.Command{
+		Use:   "test",
+		Short: "test",
+		Long: `Provisions test setup: one test user, two things and two channels. \
+						Connect both things to one of the channels, \
+						and only on thing to other channel.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			nbThings := 2
+			nbChan := 2
```
Comment (anovakovic01) [inline cli/provision.go]: Same here.

File: cli/provision.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+			ut, err := sdk.CreateToken(user)
+			if err != nil {
+				logError(err)
+				return
+			}
+
+			// Create things
+			for i := 0; i < nbThings; i++ {
+				n := "d" + strconv.Itoa(i)
```
Comment (anovakovic01) [inline cli/provision.go]: Use `fmt.Sprintf`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
@dusanb94 mnaemonics are easier to remember and dependency is minimal. Once test user is created you can reuse it for other manipulation. But I can create random string and remove dependency...

**Comment by dborovcanin** [comment]
> @dusanb94 mnaemonics are easier to remember and dependency is minimal. Once test user is created you can reuse it for other manipulation. But I can create random string and remove dependency...

I'd prefer a custom generator, but you can also consider some other packages. This one looks cool, but it's an overkill for us: https://github.com/Pallinder/go-randomdata. Anyway, since it's only for testing, please don't waste too much time on this.

File: cli/provision.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+		},
+	},
+	cobra.Command{
+		Use:   "test",
+		Short: "test",
+		Long: `Provisions test setup: one test user, two things and two channels. \
+						Connect both things to one of the channels, \
+						and only on thing to other channel.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			nbThings := 2
```
[the remark this answers is not in the document]
Comment (drasko) [reply cli/provision.go]: @anovakovic01 I have no preferences here, I can change to `num`, especially if it is used already like this in our code.

File: cli/provision.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+	"io"
+	"os"
+	"strconv"
+
+	"github.com/docker/docker/pkg/namesgenerator"
+	mfxsdk "github.com/mainflux/mainflux/sdk/go"
+	"github.com/spf13/cobra"
+)
+
+type thing struct {
```
[the remark this answers is not in the document]
Comment (drasko) [reply cli/provision.go]: Good point here.

File: cli/provision.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (anovakovic01) [reply cli/provision.go]: I think that `nb` will confuse people, that's why I would like this to be `num`.

File: cli/provision.go [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+				if err != nil {
+					logError(err)
+					return
+				}
+
+				things = append(things, m)
+			}
+			// Create channels
+			for i := 0; i < numChan; i++ {
+				n := fmt.Sprintf("c%s", strconv.Itoa(i))
```
Comment (anovakovic01) [inline cli/provision.go]: You don't need to convert, use `%d`.

File: cli/provision.go [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+			ut, err := sdk.CreateToken(user)
+			if err != nil {
+				logError(err)
+				return
+			}
+
+			// Create things
+			for i := 0; i < numThings; i++ {
+				n := fmt.Sprintf("d%s", strconv.Itoa(i))
```
Comment (anovakovic01) [inline cli/provision.go]: Same here, use `%d`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]
Just fix these two, and you have my approval.

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Pull request merged**